go/services/adr/client.go

```diff
@@ -0,0 +1,600 @@
+func (c *Client) GetAsset(
+	ctx context.Context,
+	aepName, assetName string,
+) (*Asset, error) {
+	c.logger.Debug("Getting asset", "aepName", aepName, "assetName", assetName)
+
+	req := adrbaseservice.GetAssetRequestPayload{
+		AssetName: assetName,
+	}
+
+	resp, err := c.adrBase.GetAsset(
+		ctx,
+		req,
+		protocol.WithTopicTokens{aepNameTokenKey: aepName},
+	)
+	if err != nil {
+		return nil, translateError(err)
+	}
+
+	return &resp.Payload.Asset, nil
+}
+
+// UpdateAssetStatus updates the status of an asset.
+func (c *Client) UpdateAssetStatus(
+	ctx context.Context,
+	aepName string,
+	asset *Asset,
+) (*Asset, error) {
+	c.logger.Debug(
+		"Updating asset status",
+		"aepName",
+		aepName,
+		"assetName",
+		asset.Name,
+		asset.Status,
+	)
+
+	req := adrbaseservice.UpdateAssetStatusRequestPayload{
+		AssetStatusUpdate: adrbaseservice.UpdateAssetStatusRequestSchema{
+			AssetName:   asset.Name,
+			AssetStatus: *asset.Status,
+		},
+	}
+
+	resp, err := c.adrBase.UpdateAssetStatus(
+		ctx,
+		req,
+		protocol.WithTopicTokens{aepNameTokenKey: aepName},
+	)
+	if err != nil {
+		return nil, translateError(err)
+	}
+
+	return &resp.Payload.UpdatedAsset, nil
+}
+
+// CreateDetectedAsset creates a detected asset.
+func (c *Client) CreateDetectedAsset(
+	ctx context.Context,
+	aepName string,
+	asset *DetectedAsset,
+) (*adrbaseservice.CreateDetectedAssetResponsePayload, error) {
+	req := adrbaseservice.CreateDetectedAssetRequestPayload{
+		DetectedAsset: *asset,
+	}
+
+	resp, err := c.adrBase.CreateDetectedAsset(
+		ctx,
+		req,
+		protocol.WithTopicTokens{aepNameTokenKey: aepName},
+	)
+	if err != nil {
+		return nil, translateError(err)
+	}
+
+	// Return the actual payload type instead of *NotificationResponse
+	return &resp.Payload, nil
+}
+
+// CreateDiscoveredAssetEndpointProfile creates a discovered asset endpoint profile.
+func (c *Client) CreateDiscoveredAssetEndpointProfile(
+	ctx context.Context,
+	aepName string,
+	profile *DiscoveredAssetEndpointProfile,
+) (*aeptypeservice.CreateDiscoveredAssetEndpointProfileResponsePayload, error) {
+	c.logger.Debug(
+		"Creating discovered asset endpoint profile",
+		"aepName",
+		aepName,
+	)
+
+	req := aeptypeservice.CreateDiscoveredAssetEndpointProfileRequestPayload{
+		DiscoveredAssetEndpointProfile: *profile,
+	}
+
+	resp, err := c.aepTypes.CreateDiscoveredAssetEndpointProfile(
+		ctx,
+		req,
+		protocol.WithTopicTokens{aepNameTokenKey: aepName},
+	)
+	if err != nil {
+		return nil, translateError(err)
+	}
+
+	return &resp.Payload, nil
+}
+
+// handleAssetUpdateTelemetry processes asset update telemetry messages.
+func (c *Client) handleAssetUpdateTelemetry(
+	ctx context.Context,
+	msg *protocol.TelemetryMessage[Asset],
+) error {
+	aepName := msg.TopicTokens[aepNameTokenKey]
+
+	// Skip if we are not observing the asset.
+	assetKey := aepName + "~" + msg.Payload.Name
+	c.mu.RLock()
+	_, observed := c.observedAssets[assetKey]
+	c.mu.RUnlock()
+	if !observed {
+		msg.Ack()
```

Comment:
nit: since you're always `Ack`ing anyways, it might be cleaner to `defer msg.Ack()` near the top of the method. Ditto for `handleAepUpdateTelemetry`.